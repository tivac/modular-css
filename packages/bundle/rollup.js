"use strict";

module.exports = require("modular-css-rollup");
