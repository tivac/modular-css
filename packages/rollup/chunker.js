/* eslint-disable no-loop-func */
"use strict";

const merge = (graph, original, target) => {
    const incoming = graph.incomingEdges[original];
    const outgoing = graph.outgoingEdges[original];

    graph.getNodeData(target).unshift(original);

    incoming.forEach((src) => {
        graph.removeDependency(src, original);

        if(src !== target) {
            graph.addDependency(src, target);
        }
    });

    outgoing.forEach((dest) => {
        graph.removeDependency(original, dest);

        if(dest !== target) {
            graph.addDependency(target, dest);
        }
    });

    // Bye bye
    graph.removeNode(original);
};

const setsContain = (a, b) => {
    for(const val of a) {
        if(!b.has(val)) {
            return false;
        }
    }

    return true;
};

const setsMatch = (a, b) => {
    if(a.size !== b.size) {
        return false;
    }

    return setsContain(a, b);
};

const chunks = ({ entries, graph }) => {
    const result = graph.clone();

    // Lookups
    const nodeToEntries = new Map();
    const entryToNodes = new Map();

    // Keeping track of nodes we've already handled
    const chunked = new Set();

    // First build up a map of entry -> set of deps for comparisons
    entries.forEach((entry) => {
        const deps = result.dependenciesOf(entry).reverse();

        if(!entryToNodes.has(entry)) {
            entryToNodes.set(entry, new Set());
        }

        deps.forEach((node) => {
            if(!nodeToEntries.has(node)) {
                nodeToEntries.set(node, new Set());
            }

            nodeToEntries.get(node).add(entry);
            entryToNodes.get(entry).add(node);
        });
    });

    // Keep looping until all nodes have been chunked (even if that's just into themselves)
    while(chunked.size < nodeToEntries.size) {
        // Iterate the nodes associated with each entry,
        // and figure out if the node can be collapsed
        entryToNodes.forEach((nodes, entry) => {
            // Nodes that will be merged at the end of this iteration
            const queued = new Set();

            // This will be the first set of entries found in the walk,
            // any subsequent nodes in this pass must match this to be combined
            let branches;

            // If all the nodes for this branch are handled, skip it
            if(setsContain(nodes, chunked)) {
                return;
            }

            // Walk the dependencies of the current entry in order
            const branch = result.dependenciesOf(entry).reverse();

            for(const node of branch) {
                if(chunked.has(node)) {
                    continue;
                }

                // Figure out the entries that reference this node
                const containers = nodeToEntries.get(node);

                if(!branches) {
                    branches = containers;
                } else if(!setsMatch(branches, containers)) {
                    // TODO: Can we get the order to have leaf nodes first so we could stop
                    // TODO: iteration on this branch at this point? Might have to implement BFS
                    // TODO: to get that info
                    continue;
                }

                queued.add(node);
            }

            // Merge all the queued nodes into the first node in the queue
            let base;

            queued.forEach((node) => {
                chunked.add(node);

                if(!base) {
                    base = node;

                    return;
                }

                merge(result, node, base);
            });
        });
    }

    return result;
};

module.exports = chunks;
