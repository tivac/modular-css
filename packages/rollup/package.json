{
  "name": "@modular-css/rollup",
  "version": "22.0.2",
  "description": "Add modular-css support to rollup",
  "main": "./rollup.js",
  "repository": "tivac/modular-css",
  "bugs": {
    "url": "https://github.com/tivac/modular-css/issues"
  },
  "author": "Pat Cavit <npm@patcavit.com>",
  "license": "MIT",
  "publishConfig": {
    "access": "public"
  },
  "keywords": [
    "rollup",
    "rollup-plugin",
    "css",
    "css-modules",
    "modular-css",
    "postcss"
  ],
  "dependencies": {
    "@modular-css/processor": "file:../processor",
    "dedent": "0.7.0",
    "esutils": "^2.0.2",
    "rollup-pluginutils": "^2.0.1",
    "slash": "^2.0.0"
  },
  "peerDependencies": {
    "rollup": "^1.1.2"
  }
}
