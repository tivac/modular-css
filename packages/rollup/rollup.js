/* eslint-disable max-statements, complexity */
"use strict";

const path = require("path");

const utils = require("rollup-pluginutils");
const dedent = require("dedent");
const slash = require("slash");

const Processor = require("@modular-css/processor");
const output = require("@modular-css/processor/lib/output.js");

const DEFAULT_EXT = ".css";

// sourcemaps for css-to-js don't make much sense, so always return nothing
// https://github.com/rollup/rollup/wiki/Plugins#conventions
const emptyMappings = {
    mappings : "",
};

const DEFAULTS = {
    common       : "common.css",
    dev          : false,
    json         : false,
    meta         : false,
    namedExports : true,
    styleExport  : false,
    verbose      : false,
    empties      : false,

    // Regexp to work around https://github.com/rollup/rollup-pluginutils/issues/39
    include : /\.css$/i,
};

const deconflict = (ident, existing) => {
    let proposal = ident;
    let idx = 0;

    while(existing.has(proposal)) {
        proposal = `${ident}${++idx}`;
    }

    return proposal;
};

module.exports = (opts = {}) => {
    const options = {
        __proto__ : null,
        ...DEFAULTS,
        ...opts,
    };

    const {
        dev,
        done,
        map,
        styleExport,
        verbose,
        processor = new Processor(options),
    } = options;

    const filter = utils.createFilter(options.include, options.exclude);

    // eslint-disable-next-line no-console, no-empty-function
    const log = verbose ? console.log.bind(console, "[rollup]") : () => {};

    if(typeof map === "undefined") {
        // Sourcemaps don't make much sense in styleExport mode
        // But default to true otherwise
        options.map = !styleExport;
    }

    return {
        name : "@modular-css/rollup",

        buildStart() {
            log("build start");

            // done lifecycle won't ever be called on per-component styles since
            // it only happens at bundle compilation time
            // Need to do this on buildStart so it has access to this.warn() o_O
            if(styleExport && done) {
                this.warn(
                    `Any plugins defined during the "done" lifecycle won't run when "styleExport" is set!`
                );
            }

            // Watch any files already in the procesor
            Object.keys(processor.files).forEach((file) => this.addWatchFile(file));
        },

        watchChange(file) {
            if(!processor.has(file)) {
                return;
            }

            log("file changed", file);

            // TODO: should the file be removed if it's gone?
            processor.invalidate(file);
        },

        async transform(code, id) {
            if(!filter(id)) {
                return null;
            }

            log("transform", id);

            let processed;

            try {
                processed = await processor.string(id, code);
            } catch(e) {
                // Replace the default message with the much more verbose one
                e.message = e.toString();

                return this.error(e);
            }

            const { details, exports : content } = processed;
            const { graph } = processor;

            const exportedKeys = Object.keys(content);
            const relative = path.relative(processor.options.cwd, id);
            const compositions = processor.dependencies(id, { filesOnly : false });
            const dependencies = processor.dependencies(id, { filesOnly : true });
            
            const imported = new Set();
            const defined = new Map();
            
            const out = [];

            // console.log({
            //     id,
            //     compositions,
            //     dependencies,
            // });

            // create import statements for all of the values used in compositions
            compositions.forEach((comp) => {
                const { file, selector } = graph.getNodeData(comp);

                if(!selector || file === id) {
                    return;
                }

                imported.add(selector);

                out.push(`import { ${selector} } from "${slash(file)}";`);
            });

            // Create vars representing exported @values & use them in local var definitions
            for(const key in content) {
                const classes = [];

                const selectorKey = Processor.selectorKey(id, key);
                
                if(graph.hasNode(selectorKey)) {
                    const composed = graph.dependenciesOf(selectorKey);

                    composed.forEach((dep) => {
                        const { selector } = graph.getNodeData(dep);
                        
                        if(!selector) {
                            return;
                        }

                        classes.push(selector);
                    });
                }

                // Change identifier value if it overlaps with one that was imported
                const ident = deconflict(key, imported);

                defined.set(key, ident);

                classes.push(...(Array.isArray(content[key]) ? content[key] : [ content[key] ]));
                
                out.push(`const ${ident} = ${classes.join(` + " " + `)}`);
            }

            const defaultExports = exportedKeys
                .map((key) => `${JSON.stringify(key)} : ${defined.get(key)}`)
                .join(`,\n`);

            if(dev) {
                out.push(dedent(`
                    const data = {
                        ${defaultExports}
                    };

                    export default new Proxy(data, {
                        get(tgt, key) {
                            if(key in tgt) {
                                return tgt[key];
                            }

                            throw new ReferenceError(
                                key + " is not exported by " + ${JSON.stringify(slash(relative))}
                            );
                        }
                    })
                `));
            } else {
                out.push(dedent(`
                    export default {
                        ${defaultExports}
                    };
                `));
            }

            if(options.namedExports) {
                const namedExports = exportedKeys
                    .map((key) => `${defined.get(key)} as ${key}`)
                    .join(`,\n`);

                out.push(dedent(`
                    export {
                        ${namedExports}
                    };
                `));
            }

            if(styleExport) {
                out.push(`export var styles = ${JSON.stringify(details.result.css)};`);
            }

            // Watch all the CSS files this file depends on
            dependencies.forEach((dep) => {
                this.addWatchFile(dep);
            });

            // Return JS representation to rollup
            return {
                code : out.join("\n"),
                map  : emptyMappings,

                // Disable tree-shaking for CSS modules w/o any classes to export
                // To make sure they're included in the bundle
                moduleSideEffects : exportedKeys.length || "no-treeshake",
            };
        },

        async generateBundle(outputOptions, bundle) {
            // styleExport disables all output file generation
            if(styleExport) {
                return;
            }

            const {
                file,
                dir,
                assetFileNames,
            } = outputOptions;

            const chunks = new Map();
            const used = new Set();
            const unused = [];

            // Determine the correct to option for PostCSS by doing a bit of a dance
            const to = (!file && !dir) ?
                path.join(processor.options.cwd, assetFileNames) :
                path.join(
                    dir ? dir : path.dirname(file),
                    assetFileNames
                );
            
            // Walk bundle, create CSS output files
            Object.keys(bundle).forEach((entry) => {
                const { type, modules, name } = bundle[entry];
                
                /* istanbul ignore if */
                if(type === "asset") {
                    return;
                }

                const deps = Object.keys(modules).reduce((acc, f) => {
                    if(processor.has(f)) {
                        const css = processor.normalize(f);
                        
                        used.add(css);
                        acc.push(css);
                    }

                    return acc;
                }, []);

                if(!deps.length) {
                    return;
                }

                chunks.set(entry, { deps, name });
            });

            // Add any bare CSS files to be output as part of the common chunk
            processor.dependencies().forEach((css) => {
                if(used.has(css)) {
                    return;
                }

                unused.push(css);
            });

            if(unused.length) {
                const { name } = path.parse(options.common);

                chunks.set("common", { deps : unused, name });
            }

            // If assets are being hashed then the automatic annotation has to be disabled
            // because it won't include the hashed value and will lead to badness
            let mapOpt = map;

            if(assetFileNames.includes("[hash]") && typeof mapOpt === "object") {
                mapOpt = {
                    __proto__  : null,
                    ...mapOpt,
                    annotation : false,
                };
            }

            // Track specified name -> output name for writing out metadata later
            const names = new Map();

            // Track chunks that don't actually need to be output
            const duds = new Set();

            for(const [ entry, { deps, name }] of chunks) {
                /* eslint-disable-next-line no-await-in-loop */
                const result = await processor.output({
                    // Can't use this.getAssetFileName() here, because the source hasn't been set yet
                    //  Have to do our best to come up with a valid final location though...
                    to : to.replace(/\[(name|extname)\]/g,  (match, field) =>
                        (field === "name" ? name : DEFAULT_EXT)
                    ),
                    map : mapOpt,

                    files : deps,
                });

                // Don't output empty files if empties is falsey
                if(!options.empties && !result.css.length) {
                    duds.add(entry);

                    continue;
                }

                const id = this.emitFile({
                    type   : "asset",
                    name   : `${name}${DEFAULT_EXT}`,
                    source : result.css,
                });

                // Save off the final name of this asset for later use
                const dest = this.getFileName(id);

                names.set(entry, dest);

                log("css output", dest);

                if(result.map) {
                    // Make sure to use the rollup name as the base, otherwise it won't
                    // automatically handle duplicate names correctly
                    const fileName = dest.replace(DEFAULT_EXT, `${DEFAULT_EXT}.map`);

                    log("map output", fileName);

                    this.emitFile({
                        type   : "asset",
                        source : result.map.toString(),

                        // Use fileName instead of name because this has to follow the parent
                        // file naming and can't be double-hashed
                        fileName,
                    });

                    // Had to re-add the map annotation to the end of the source files
                    // if the filename had a hash, since we stripped it out up above
                    if(assetFileNames.includes("hash")) {
                        bundle[dest].source += `\n/*# sourceMappingURL=${path.basename(fileName)} */`;
                    }
                }
            }

            if(options.json) {
                const dest = typeof options.json === "string" ? options.json : "exports.json";

                log("json output", dest);

                const compositions = await processor.compositions;

                this.emitFile({
                    type   : "asset",
                    name   : dest,
                    source : JSON.stringify(compositions, null, 4),
                });
            }

            if(options.meta) {
                const meta = {};

                for(const [ entry ] of chunks) {
                    const chunk = bundle[entry];

                    if(!chunk) {
                        continue;
                    }

                    // Attach info about this asset to the bundle
                    const { assets = [] } = chunk;

                    assets.push(names.get(entry));

                    chunk.assets = assets;

                    meta[entry] = {
                        assets,
                    };
                }
            
                const dest = typeof options.meta === "string" ? options.meta : "metadata.json";

                log("metadata output", dest);

                this.emitFile({
                    type   : "asset",
                    source : JSON.stringify(meta, null, 4),
                    name   : dest,
                });
            }
        },
    };
};
