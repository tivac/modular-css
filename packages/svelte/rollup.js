"use strict";

const fs = require("fs");
const path = require("path");

const mkdirp = require("mkdirp");

const Processor = require("modular-css-core");
const methods = require("./methods.js");

module.exports = function(args) {
    const processor = new Processor(args);
    let source;

    return {
        preprocess : {
            markup : methods.markup(processor),
            style  : methods.style
        },

        plugin : {
            name : "modular-css-rollup-svelte",

            options : ({ input }) => {
                source = input;
            },

            ongenerate : (bundle, result) => {
                result.css = processor.output({
                    from : source,
                    to   : args.css
                });
            },

            onwrite : (bundle, result) =>
                result.css.then((data) => {
                    if(args.css) {
                        mkdirp.sync(path.dirname(args.css));
                        
                        fs.writeFileSync(
                            args.css,
                            data.css
                        );
                    }
                })
        }
    };
};
