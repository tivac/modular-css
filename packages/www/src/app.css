body {
    font-family: 'Roboto', sans-serif;
}

pre,
textarea,
:global(.CodeMirror) {
    font-family: 'Inconsolata', monospace;
}

.page {}

.input {}

.tabs {}

.tab {}

.tabon {
    composes: tab;
}

.add {}

.file {}

.remove {}

.output {}

.error {}

.erroron {}

.result {}

.compositions {}

