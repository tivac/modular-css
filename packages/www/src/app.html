<!DOCTYPE html>
<html lang="en">
<head>
	<meta charset="utf-8" />
	<meta name="description" content="" />
	<meta name="viewport" content="width=device-width, initial-scale=1" />
	<link href="https://unpkg.com/normalize.css@8.0.1/normalize.css" rel="stylesheet" />
	<link href="https://fonts.googleapis.com/css?family=Roboto" rel="stylesheet" />
	%svelte.head%
</head>
<body>
	%svelte.body%
</body>
</html>
