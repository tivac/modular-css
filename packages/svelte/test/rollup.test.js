/* eslint consistent-return: off */
"use strict";

const path = require("path");

const shell = require("shelljs");
const { rollup } = require("rollup");

const write = require("@modular-css/test-utils/write.js")(__dirname);
const prefix = require("@modular-css/test-utils/prefix.js")(__dirname);
const dir = require("@modular-css/test-utils/read-dir.js")(__dirname);
const watching = require("@modular-css/test-utils/rollup-watching.js");

const plugin = require("../svelte.js");

const assetFileNames = "assets/[name][extname]";
const format = "es";

describe("/svelte.js", () => {
    let warnSpy;

    beforeEach(() => {
        warnSpy = jest.spyOn(global.console, "warn");
        warnSpy.mockImplementation(() => { /* NO-OP */ });
    });

    describe("rollup watching", () => {
        const { watch } = require("rollup");
        let watcher;

        beforeAll(() => shell.rm("-rf", prefix(`./output/rollup/*`)));
        afterEach(() => watcher.close());

        it("should generate updated output", async () => {
            const { preprocess, processor } = plugin();

            // Create v1 of the files
            write(`./rollup/input/index.js`, `
                import app from "./app.html";
                console.log(app);
            `);

            write(`./rollup/input/app.html`, `
                <div class="{css.nope}">Hi</div>
            `);

            // Start watching
            watcher = watch({
                input  : prefix(`./output/rollup/input/index.js`),
                output : {
                    file : prefix(`./output/rollup/output/output.js`),
                    format,
                    assetFileNames,
                },
                plugins : [
                    require("rollup-plugin-svelte")({
                        preprocess,
                    }),
                    require("@modular-css/rollup")({
                        processor,
                    }),
                ],
            });

            const wait = watching.promise(watcher);

            await wait();

            const v1 = dir("./rollup/output/");

            setTimeout(() => {
                write(`./rollup/input/app.css`, `
                    .nope {
                        color: blue;
                    }
                `);

                write(`./rollup/input/app.html`, `
                    <link rel="stylesheet" href="./app.css" />
                    <div class="{css.nope}">Hi</div>
                `);
            }, 100);

            await wait();

            const v2 = dir("./rollup/output/");

            expect(v1).toMatchDiffSnapshot(v2, {
                // Get specific to avoid some travis-related weirdness
                contextLines     : 0,
                stablePatchmarks : true,
            });
        });

        it("should generate updated output when composition changes", async () => {
            const { preprocess, processor } = plugin();

            // Create v1 of the files
            write(`./rollup-composes/input/index.js`, `
                import app from "./app.html";
                console.log(app);
            `);

            write(`./rollup-composes/input/app.html`, `
                <link rel="stylesheet" href="./app.css" />
                <div class="{css.a}">Hi</div>
            `);

            write(`./rollup-composes/input/app.css`, `
                .a {
                    composes: b from "./other.css";

                    color: red;
                }
            `);
            
            write(`./rollup-composes/input/other.css`, `
                .b {
                    background: blue;
                }

                .c {
                    background: green;
                }
            `);

            // Start watching
            watcher = watch({
                input  : prefix(`./output/rollup-composes/input/index.js`),
                output : {
                    file : prefix(`./output/rollup-composes/output/output.js`),
                    format,
                    assetFileNames,
                },
                plugins : [
                    require("rollup-plugin-svelte")({
                        preprocess,
                    }),
                    require("@modular-css/rollup")({
                        processor,
                    }),
                ],
            });

            const wait = watching.promise(watcher);

            await wait();

            const v1 = dir("./rollup-composes/output/");

            setTimeout(() => write(`./rollup-composes/input/app.css`, `
                .a {
                    composes: c from "./other.css";
                    
                    color: red;
                }
            `), 100);
            
            await wait();

            const v2 = dir("./rollup-composes/output/");

            expect(v1).toMatchDiffSnapshot(v2);

            wait.close();
        });
    });

    describe("rollup errors", () => {
        it.each([[ "link" ], [ "style" ]])(
            "should show useful errors from rollup (<%s>)",
            async (type) => {
                const { preprocess, processor } = plugin();

                try {
                    await rollup({
                        input : "./error.js",

                        plugins : [
                            require("rollup-plugin-hypothetical")({
                                cwd   : path.join(__dirname, "./specimens"),
                                files : {
                                    "./error.js" : `
                                        import Component from "./error-${type}.html";

                                        console.log(Component);
                                    `,
                                },

                                allowFallthrough : true,
                            }),
                            require("rollup-plugin-svelte")({
                                preprocess,
                            }),
                            require("@modular-css/rollup")({
                                processor,
                            }),
                        ],
                    });
                } catch(e) {
                    expect(e.toString()).toMatch(/\.wooga/);
                }
            }
        );
    });

    describe("rollup chunking", () => {
        it("should correctly chunk svelte files using inline <style>", async () => {
            const { preprocess, processor } = plugin();

            const bundle = await rollup({
                input : [
                    require.resolve("./specimens/inline-chunking/a.html"),
                    require.resolve("./specimens/inline-chunking/b.html"),
                ],

                plugins : [
                    require("rollup-plugin-svelte")({
                        preprocess,
                    }),
                    require("@modular-css/rollup")({
                        processor,
                    }),
                ],
            });

            const out = await bundle.generate({
                format         : "esm",
                assetFileNames : "[name][extname]",
                chunkFileNames : "[name]",
            });

            expect(out).toMatchRollupAssetSnapshot();
        });
    });
});
