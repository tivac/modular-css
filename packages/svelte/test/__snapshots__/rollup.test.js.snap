// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/svelte.js rollup chunking should correctly chunk svelte files using inline <style> 1`] = `
Object {
  "a.css": "
/* packages/svelte/test/specimens/inline-chunking/a-dep.css */
.mc6ca450b0_adep {
    margin: 5px;
}
/* packages/svelte/test/specimens/inline-chunking/a.html */
.mc80aa5044_a {
    color: aqua;
}
",
  "b.css": "
/* packages/svelte/test/specimens/inline-chunking/b-dep.css */
.mcf56d52f6_bdep {
    padding: 5px;
}
/* packages/svelte/test/specimens/inline-chunking/b.html */
.mcbb945b2f_b {
    color: blue;
}
",
  "shared.css": "
/* packages/svelte/test/specimens/inline-chunking/shared.css */
.mcfef68931_shared {
    background: black;
}
",
}
`;

exports[`/svelte.js rollup watching should generate updated output 1`] = `
Snapshot Diff:
- First value
+ Second value

@@ --- --- @@
+   "assets/app.css": "/* packages/svelte/test/output/rollup/input/app.css */
+ .mc043115d5_nope {
+     color: blue;
+ }",
@@ --- --- @@
+ 
@@ --- --- @@
+ 
@@ --- --- @@
- 			div.className = \\"css.nope\\";
+ 			div.className = \\"mc043115d5_nope\\";
`;

exports[`/svelte.js rollup watching should generate updated output when composition changes 1`] = `
Snapshot Diff:
- First value
+ Second value

@@ --- --- @@
  
  	return {
  		c() {
  			div = createElement(\\"div\\");
  			div.textContent = \\"Hi\\";
- 			div.className = \\"mc8c3b9f40_b mc7c0d2c82_a\\";
+ 			div.className = \\"mc8c3b9f40_c mc7c0d2c82_a\\";
  		},
  
  		m(target, anchor) {
  			insert(target, div, anchor);
  		},
`;
